src/main.rs

```diff
@@ -73,13 +74,18 @@ fn interp(prog: &[u8], map: Vec<usize>) {
         stdout: std::io::stdout(),
     };
 
-    // FIXME only use locations for loop heads.
-    let mut locs = Vec::with_capacity(icx.prog.len());
-    locs.resize_with(icx.prog.len(), Location::new);
+    // Locations are lazily instantiated later as loop-headers are found.
+    let mut locs: HashMap<usize, Location<_>> = HashMap::new();
```

Comment:
This completely negates all our careful work in `Location` to avoid hashmaps :p Remember that `Location` is just a `usize`, so a Vec of them is really cheap, even if most are unused. It'll be *way* faster than this mechanism.

Reply:
We can easily make this a `Vec::with_capacity(prog.len())`.

Reply:
See https://github.com/softdevteam/ykbf/pull/9#discussion_r570280514.

I can do a vec now if we wish, but this seems secondary to what I'm doing here in this PR.

Reply:
To be clear, we are OK with storing `None` if a location isn't a loop header?

Reply:
Why would you store `None`? `Option<Location>` will be two machine words big! It's fine not to use a `Location`: we implicitly guarantee that, if you don't use it, it will consume precisely 1 machine word, no more no less.

Reply:
How do you tell the control point that the location you passed is not to be used?

Reply:
I don't understand. How to you express that a bytecode is not a loop header?

Reply:
We can squash the representation into 1 machine word by storing an index into an array of type `Location` where the index `usize::MAX` means "not a loop", but you will always need some special value to represent "not a loop".

Reply:
> Edit: I think I see now: If the bytecode is not [ then you just return None, without looking it up in the vec.

Yes, that's what I'm assuming we can do. I might be wrong though!

Reply:
> Edit: I think I see now: If the bytecode is not [ then you just return None, without looking it up in the vec.

But the vector can't be sparse :P

Reply:
Line 82 already contains the necessary logic doesn't it?

Reply:
Line 82 checks if the bytecode at icx.pc is a loop header. The next step is to find (or create) the location for this program point, which currently involves looking into a hashmap, which is sparse.

Suppose the hashmap is {1 => l1, 4 => l2}. How do you represent the fact that program points 0, 2, and 3 are not loop headers?

One way is `vec![None, Some(l1), None, None, Some(l2)]`.

Another is `vec![l1, l2]` and each bytecode stores an index into this array, and some arbitrary index (say `usize::MAX`) means that the bytecode doesn't have a `Location`.
If you want to convert that to a vector, then how do you deal with a program point which doesn't need a location?

Reply:
There's no advantage to having a `None` over a `Location` that you never actually use.

Reply:
Are you saying, use `vec![Location::new(); prog.len()]`?

If we do this then doesn't it mean that we will profile and potentially compile traces starting from program points which are not the beginning of a loop?

Reply:
> Are you saying, use vec![Location::new(); prog.len()]?

Yes.

> If we do this then doesn't it mean that we will profile and potentially compile traces starting from program points which are not the beginning of a loop?

You could put cabbage inside that vec; it only becomes a problem if you pass the cabbage to the control point. In other words line 82 can say "if a loop header, pass the corresponding location in the list; otherwise pass `None`".

Reply:
Wait, I think I see what you mean, but it's kind of odd... Let me write the code...

Reply:
I think you mean like 4006da1?

I find this quite an odd way to do things. I'd prefer it if by construction program points that are not the start of a loop had no corresponding `Location` at all. If only for code comprehension.

At the same time I do want to move on....

@ptersilie what do you think?

Reply:
Yep -- and I don't think it's too odd ;)

Reply:
Fixed in 4006da1.

Reply:
You've already pointed to that commit?

Reply:
I think you meant https://github.com/softdevteam/ykbf/pull/9/commits/184ab2a12b6e8bbcb4d3c35d5a6d2ed8d168426c.

Reply:
That matches my understanding of what Laurie meant earlier. I think it's fine. Maybe add a comment saying that not all Locations will be used since we are only interested in the ones that start at `[`, and that storing `Location`s is better than storing `Option`s.

Reply:
How about 19b7509?